_posts/en/newsletters/2019-05-28-newsletter.md

```diff
@@ -0,0 +1,597 @@
+outputs matched the hash digest read from the script by COSHV.
+
+Comparing this to our example above, Alice would again ask each of the
+participants for a public key (such as a Taproot
+address[^taproot-pubkeys]).  Similar to before, she'd create 10 outputs
+which each paid one of the receivers---but she wouldn't need to form
+this into a specific distribution transaction.  Instead, she'd just hash
+the ten outputs together and use the resultant digest to create a
+tapleaf script containing COSHV.  That would be the only tapleaf in this
+Taproot commitment.  Alice could also use the participants' public keys
+to form the taproot internal key to allow them to cooperatively spend
+the money without revealing the Taproot script path.
+
+![FIXME](/img/posts/2019-05-tx-tree-2.png)
+
+Alice could then give each of the receivers a copy of all ten outputs to
+allow each of the receivers to verify that Alice's setup transaction,
+when suitably confirmed, guaranteed them payment.  When they later
+wanted to spend that payment, any of them could then create a distribution
+transaction containing the committed outputs.  Unlike the example from
+the previous subsection, they would never need to interact with each
+other.  Even better, the information Alice needs to send them in order
+to allow them to verify the setup transaction and ultimately spend their
+money could be sent through existing asynchronous communication methods
+such as email or a cloud drive.  That means the receivers wouldn't need
+to be online at the time Alice created and sent her setup transaction.
+
+This elimination of the need to interact is a particular highlight of
+the proposal.  If we imagine the example above with Alice being an
+exchange, the interactive form of the protocol would require her to keep
+the ten participants online and connected to her service from the moment
+each of them submitted their withdrawal request until the interaction
+was done---and they'd all need to use wallets compatible with such a
+child transaction signing protocol.  The non-interactive form with COSHV
+would only require them to submit a Bitcoin address and an email address
+(or some other protocol address for delivery of the committed outputs).
+
+### More advanced usage
+
+The transaction size (and thus fee cost) of using COSHV in the previous
+subsection's example is basically the same as using the pre-signed
+transaction format with current Bitcoin transactions.  However, the
+ability to use Taproot to allow the receivers to choose from any one of
+up to several billion spending conditions adds a significant amount of
+flexibility to the approach.
+
+In our previous examples, getting the distribution transaction confirmed
+was an all-or-nothing endeavor.  Either all ten receivers finally
+received their money at the same time, or all of them were still waiting
+for that confirmation to occur.  However, with a MAST tree and COSHV,
```

Comment:
I think the approach Jeremy is thinking of is a tree of outputs, ie:

 * I want to pay to 125 people
 * I make 25 COSHV outputs of 5 payments each
 * I make 5 second level COSHV outputs, each grouping 5 of the 25 COSHV outputs I just generated
 * I make a transaction that pays to a COSHV commitment of those 5 COSHV outputs

If one of the 125 people wants to collect their funds, they publish three transactions: one that expands the single COSHV commitment into 5; another that expands one of those into another 5; and a final one that expands one of those into 5 actual payments, including their own.

I think the more interesting (and scalable) MAST approach is to allow increasing fees: if Alice is paying 3 people, create a MAST tree of:

     [1 Bob, 2 Carol, 1.5 Dave, 0.4 Alice] OP_COSHV
     100 OP_CSV [1 Bob, 2 Carol, 1.5 Dave, 0.3995 Alice] OP_COSHV
     2000 OP_CSV [1 Bob, 2 Carol, 1.5 Dave, 0.39 Alice] OP_COSHV

allowing the first path to be used immediately, but adding 0.0005 BTC in fees via the second path if there hasn't been a low fee period within 100 blocks, and adding 0.01 BTC in fees if there hasn't been a low enough fee period in two weeks.